\PassOptionsToPackage{utf8}{inputenc}
\documentclass{bioinfo}

\usepackage[draft]{hyperref}
\usepackage{makecell}
\usepackage{comment}
\usepackage{subcaption}
\usepackage{algorithm2e}

\SetAlgoLined
\SetKwProg{MyStruct}{Struct}{ contains}{end}

\newcommand{\vocab}{\textbf}

\copyrightyear{XXXX} \pubyear{XXXX}

\access{Advance Access Publication Date: Day Month Year}
\appnotes{Genome Analysis}

\begin{document}
\firstpage{1}

\subtitle{Genome Analysis}

\title[short Title]{ODGI: understanding pangenome graphs}
\author[Guarracino, Heumos \textit{et~al}.]{
Andrea Guarracino\,$^{\text{\sfb 1} \dagger}$,
Simon Heumos\,$^{\text{\sfb 2} \dagger}$,
Sven Nahnsen\,$^{\text{\sfb 2}}$,
Pjotr Prins\,$^{\text{\sfb 3}}$,
and Erik Garrison\,$^{\text{\sfb 3}*}$
}

\address{
$^{\text{\sf 1}}$University of Tor Vergata, Rome, Italy \\
$^{\text{\sf 2}}$Quantitative Biology Center (QBiC), University of T\"ubingen, T\"ubingen, Germany, 72076 \\
$^{\text{\sf 3}}$University of Tennessee Health Science Center, Memphis, TN, USA
}

\corresp{$^\ast$To whom correspondence should be addressed. \\
$^\dagger$Contributed equally.}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

% XXX key message of the paper is that we have collected a set of algorithms that enable easy use of pangenome graphs for investigating biology

\abstract{
\textbf{Motivation:} Pangenome graphs provide a complete representation of the mutual alignment of collections of genomes.
These models offer the opportunity to study the total genomic diversity of a population, including structurally complex regions.
Analyzing hundreds of gigabase-scale genomes using this model is difficult and not well-supported by existing tools.
%In addition, vertebrate genomes present highly repetitive regions which increase the complexity of the analysis.
Hence, fast and versatile software is required to ask intricate questions of such data in an efficient way. \\
\textbf{Results:} Here we present ODGI, a new suite of tools that implements scalable algorithms and has an efficient in-memory representation of DNA variation graphs. ODGI includes tools for detecting complex regions, extracting loci, removing artifacts, exploratory analysis, manipulation, validation, and visualization. Its fast parallel execution facilitates routine pangenomic tasks as well as pipelines that can quickly answer complex biological questions of gigabase-scale pangenome graphs. \\
%ODGI supports graphs in the Graphical Fragment Assembly format version 1 (GFAv1).
%The tools scale efficiently to large collections of eukaryotic genomes, allowing users to work easily with pangenome graphs of tens to thousands of large genomes while maintaining fast execution runtime and low memory overhead.
\textbf{Availability:} ODGI is under the MIT license. Source code can be downloaded from \url{https://github.com/pangenome/odgi} while the documentation is available at \url{https://odgi.readthedocs.io}.
ODGI can be installed via Bioconda \url{https://bioconda.github.io/recipes/odgi/README.html} or Guix \url{https://github.com/ekg/guix-genomics/blob/master/odgi.scm}. \\
\textbf{Contact:} \href{egarris5@uthsc.edu}{egarris5@uthsc.edu} \\
\textbf{Supplementary information:} Supplementary data are available at \textit{Bioinformatics} online.
}

\maketitle

\section{Introduction}

A pangenome models the full set of genomic elements in a given species or clade \citep{cpang2018,Eizenga_2020}.
In contrast to reference-based approaches which relate sequences to a single genome, these data structures encode the mutual relationships between all the genomes represented.
In pangenome graphs \citep{Paten:2017}, homologous regions between genomes are compressed into a single representative model of all alleles present in the pangenome.
These flexible models let us encode any kind of variation, allowing the generation of truly unbiased data systems on which to base analyses of genome evolution.
Although these structures are of clear utility to researchers \citep{cpang2018}, and have been the focus of numerous applications in read mapping \citep{Garrison:2018,Baaijens_2019,Hickey:2020,Sibbesen_2021}, we observe that the community lacks a toolset specifically focused on graph manipulation and interrogation.

Projects like the Human Pangenome Reference Consortium (HPRC) and Telomere-to-Telomere (T2T) consortium \citep{Miga:2020, Logsdon_2021, Nurk_2021} have recently demonstrated that high-quality \textit{de novo} assemblies can be routinely generated from third-generation long read sequencing data.
We anticipate that in the near future, studies will frequently begin from assemblies of similar quality, leading to demand for methods that allow us to understand pangenomes.
As experience in the HPRC has shown, pangenome graphs are a natural model to support these kinds of analyses, but studies based on pangenome graphs are difficult due to a lack of downstream tools and standard operations.

Here, we present the Optimized Dynamic Genome/Graph Implementation (ODGI) toolkit, a genome graph interrogation and transformation system specifically built to handle the data scales encountered when working with pangenomes built from hundreds of haplotype-resolved genomes.
ODGI provides a set of standard operations on the pangenome variation graph data model, generalizing ``genome arithmetic'' concepts like those found in BEDTools \citep{Quinlan_2010} to work on genome graphs, and providing new operations, such as visualizations, sorting, and liftover projections critical to understanding and exploiting pangenome graphs.

%Tools in ODGI operate on an efficient dynamic HandleGraph model \citep{Eizenga_2020_BX}.
%Algorithms written against this abstract API can be applied to the graph.% applies algorithms based on the HandleGraph API.
%This common API lets us reuse and and extend algorithms shared with the VG toolkit (VG) \citep{Garrison:2018}.
%We specifically develop new methodsfocused on problems encountered when building pangenome graphs at the scale of vertebrate populations.
%To ease interactive use, the majority of the ODGI tools are implemented in an index-free manner, avoiding the need to create index structures at each step of complex graph processing pipelines.
%Thanks to ODGI's efficient path representation, its tools can work with variation graphs with highly complex regions.
%This eliminates one of the major bottlenecks when working with very large and deep variation graphs, and allows researchers to build and understand graphs of previously-inaccessible complexity and scale.


%\footnote{\url{https://humanpangenome.org/} (accessed Oct 2021)}
%\footnote{\url{https://sites.google.com/ucsc.edu/t2tworkinggroup/home} (accessed Oct 2021)}.

\section{Model}

A pangenome graph is a sequence model that encodes the mutual alignment of many genomes \citep{Garrison_2019_thesis,Eizenga_2020}.
In the variation graph, $V = (N, E, P)$, nodes $N = n_1\ldots n_{|N|}$ contain sequences of DNA.
Each node $n_i$ has an identifier $i$ and an implicit reverse complement $\bar{n_i}$, and a node strand $s$ corresponds to one such orientation.
Edges $E = e_1\ldots e_{|E|}$ represent ordered pairs of node strands: $e_i = ( s_a, s_b )$.
Paths $P = p_1\ldots p_{|P|}$ describe walks over node strands: $p_i = s_1 \ldots s_{|p_i|}$.
When used as a pangenome graph, $V$ expresses sequences, haplotypes, and annotations as paths.
%The utility of the variation graph model lies in its lossless representation of genomes and their alignment.
By containing both the sequences and information about their relative variations, the variation graph provides a powerful foundation for many bioinformatic applications.

Such a graph can be constructed by a process like multiple sequence alignment \citep{Lee_2002,Grasso_2004}, but this partially-ordered graph cannot represent complex variation like inversions and cycles that occur at structural scales in genomes.
General approaches to building pangenome graphs transitively reduce an alignment between sequences to an equivalent, labeled sequence graph \citep{Kehr_2014,Garrison_2019_thesis}.
Current methods to build these graphs are still in development \citep{Li:2020,Armstrong:2020,pggb}, but they have largely settled on a common data model, represented in the Graphical Fragment Assembly (GFA) format \citep{GFA}).
%The improvement of these methods depends on tools that allow us to investigate 
This standardization supports the development of a reference set of tools that operate on the pangenome graph model, while the need to understand the graphs made by these methods motivates it.

%The improvement of these approaches depends on scalable tools for understanding pangenome graphs, a need that our work responds to.

%As a toolkit for understanding these graphs, ODGI 

%A more general approach is to build the graph from an alignment between sequences.
%We first transitively collapse characters in the input sequences that align together into single character in the output genome
%By projecting input sequences through this transformation, we obtain paths that losslessly encode the original sequences, but in the space of the graph \citep{Garrison_2019_thesis}.
%Methods to build such graphs are in active development, and here we focus on graphs built with the PanGenome Graph Builder pipeline \citep{pggb}, a process that converts an input pangenome (in FASTA) into a combined graphical model (in the Graphical Fragment Assembly (GFA) format \citep{GFA}).

\section{Implementation}

%ODGI is designed to build and modify paths in parallel, keeping its representation in memory efficient.
The ODGI toolkit builds on existing approaches to efficiently store and manipulate variation graphs \citep{Garrison:2018}.
%It implements the \textsc{HandleGraph} model \cite{Eizenga_2020_BX}.
Like other efficient libaries presenting the \textsc{HandleGraph} model \citep{Eizenga_2020_BX}, its implementation rests on three key properties which hold for most pangenome variation graphs:

\begin{enumerate}
\item They are relatively sparse, with low average node degree.
\item They can be sorted so that most edges go between nodes that are close together in the sort order.
\item Paths are locally similar to each other.
%We incorporate concepts first introduced in the dynamic version of the Graph BWT \citep{Siren:2020} (GBWT), a scalable implementation of the graph extension of the positional Burrows-Wheeler transform \citep{Durbin_2014}.
\end{enumerate}

These properties have already been used to build efficient dynamic variation graph data structures \citep{Siren:2020,Eizenga_2020_BX}.
Sparsity (1) allows us to encode edges $E$ using adjacency lists rather than matricies or hash tables.
The local linear structure of the graph (2) lets us assign node identifiers that increase along the linear components of the graph, which supports a compact storage of edges and path steps as relativistic (usually small) differences rather than absolute (always large) integer identifiers.
Path similarity (3) allows us to write local compressors that reduce the storage cost of collections of path steps.

%Here, we build on this work and focus on issues that have arisen during our recent work on the Human Pangenome.
Here, we improve on these prior efforts, focusing on issues that arose during our work with high-quality assemblies that cover almost all parts of the genome \citep{Logsdon_2021,Nurk_2021}.
%A key design objective, motivated by experience working with graphs built from high quality eukaryotic genome assemblies, is to support graphs with very high path depth.
We find that it is necessary to support graphs that have regions of very high path depth.
Such motifs can occur in collapsed repeat structures generated by ambiguous sequence homology relationships in repeats found in the centromeres and other segmental duplications.
If we cannot process such regions, we have only two options---we can remove these regions, or we can leave them unaligned.
Neither of these solutions allows us to investigate their biological features.
To seamlessy represent such difficult regions, we follow the approach of the dynamic Graph BWT (GBWT) \citep{Siren:2020}, and built a node-centric, dynamic, compressed model of the paths.
This design supports node-local modification and update of the graph, which lets us build and transform the model in parallel.

We store the graph in a vector of node structures, each of which presents a node-local view of the graph sequence, topology, and path layout.
Expressed in terms of the variation graph $V$, ODGI's core $Node$ structure includes.



\begin{algorithm}
\MyStruct{Node}{
    \textbf{id} $\in \mathbb{N}$ \tcp{an identifier}
    \textbf{lock} \tcp{atomic locking primitive}
    \textbf{sequence} $= [ATGC]*$ \tcp{DNA, RNA, protein}
    \tcp{succinct dynamic bitvector of edges}
    \textbf{edges} $= (s_{x},s_{y})* : x \land y \in$ decoding \\
    \tcp{a vector of id deltas}
    \textbf{decoding} $= (1 \ldots x \ldots N)*$ \\
    \tcp{a vector of path steps}
    \textbf{paths}
}
\caption{ODGI's relativistically-packed node structure.}
\end{algorithm}

\begin{comment}
TODO:
    struct step_t {
        uint64_t path_id;
        bool is_rev;
        bool is_start;
        bool is_end;
        uint64_t prev_id;
        uint64_t prev_rank;
        uint64_t next_id;
        uint64_t next_rank;
    };
\end{comment}

%\begin{enumerate}
%\item a node sequence $seq(n_i)$
%\end{enumerate}


Each structure contains the sequence of the node, its edges in both directions (forward and backward), and a vector of path steps that describes the previous and next steps in paths that walk across the node.
As in the GBWT, to compress the encoding, we encode path steps using a local alphabet that maps the N neighbours of the node into the range $1\ldots N$.
To further save space, node deltas, rather than IDs, are stored in this alphabet.
The delta between two nodes is defined as their distance in the graph vector (i.e., the difference between the node offsets).
The path step vector, of length n, is stored using a dynamic succinct integer vector that requires O(5nw) bits, where w is ~log2(N).
Finally, to allow operating on path steps in parallel, each node structure includes a byte-width mutex lock.

ODGI is entirely focused on “first-order” operations on the graph (Table 1), providing, unlike VG, no method to construct graphs from raw sequences or map reads.
It supports interrogation of the graph for basic dimensional properties, positional and path-related queries, and visualization in one-dimension (1D) and two-dimensions (2D).
ODGI also supports operations on the graph such as subset, subdivide, break, combine, normalize, or order its components, nodes, and paths.
Most of the tools are designed to be applied together, piping the output of one tool into the next, thereby preventing the creation of intermediate files, and reducing the number of IO operations.



\begin{comment}
%\subsection{Core functionality}

Variation graphs are commonly represented by a subset of version 1 of the Graphical Fragment Assembly format (GFAv1): odgi build and odgi view allow to convert graphs from GFA to ODGI format and vice versa, respectively.
In the variation graph model, paths have to respect the graph’s topology: this can be verified with odgi validate, to ensure no errors in the input or edited graphs.

Applying odgi stats, users can retrieve metrics describing the graph properties, such as the number of nodes, edges, paths, and graph length.
ODGI also offers more advanced tools for interrogating the graphs. odgi bin summarizes the path information into bins of specified size, enabling a summarized view of gigabase scale graphs.
odgi depth returns the node depth as defined by query criteria, allowing users to detect the complex regions in the graph due to sequences with highly identical repeats.
Complex motifs can be also detected with odgi degree, which returns the node degree as defined by query criteria.
High degree nodes are the mirror of misassemblies or problems in the pangenome building, making the tool useful for debugging the process as well.

ODGI allows edit operations on the graphs.
Cycles in the graph complicate downstream analyses: odgi break removes the cycles, reducing the complexity of the graph topology.
odgi groom removes spurious inverting links by exploring the graph from the orientation supported by most paths.
To enable efficient sequence alignment against the graph, long nodes can be divided into shorter nodes at a maximum requested size using odgi chop.
Partial order alignment, which consists of aligning sequences against a directed acyclic graph (DAG), is frequently used in pangenome building pipelines, but the current implementations return DAGs with 1-bp long nodes; odgi unchop allows joining nodes that can be merged without changing the graph topology, nor the embedded sequences, obtaining an equivalent, but more compact, representation of the graph.

Pangenome graphs can embed multiple chromosomes as separated connected components (inter-chromosomal structural variants would join the components into bigger ones).
odgi explode separates the connected components in different ODGI format files, while odgi squeeze allows merging multiple graphs into the same ODGI format file, preventing node ID collisions.
odgi extract allows extracting specific subregions of the graph as defined by query criteria, thereby simplifying the downstream analyses and reducing the resources to work only with the extracted region.

In variation graphs the coordinates are provided by the embedded path sequences.
Indeed, the node IDs are not meant to be stable. odgi position finds, translates, and liftovers graph and path positions between different graphs by exploiting their shared path sequences (Figure 1.B).
\end{comment}

\begin{comment}
key message of the paper is that we have collected a set of algorithms that enable easy use of pangenome graphs for investigating biology
-> build model solves problem of working with big graphs in memory
-> view (convert to GFA) & paths solve problem of exporting basic features of the graph (e.g. paths)
-> stats (understand basic size / structure) & bin & degree & depth solves problem of understanding the overall structure and size of the graph
-> sort (groom) & layout solves problem of finding latent structure in the pangenome
-> viz & draw provides a human-viewable readout of the graph
-> chop & unchop & squeeze & break & prune & explode lets us break apart or combine the graph nodes and topology
-> position & tips & untangle (jaccard based coordinate conversion) provides a way to map coordinates between any genomes in the graph (e.g. liftover!)
-> extract lets us pull out specific regions of the graph based on path ranges, nodes and positions
\end{comment}

\section{Results}

\subsection{Building the \textsc{ODGI} model}

% erik

% build
Solves problem of working with big and complex graphs in limited memory.

Figure: encoding model as a cartoon --  HTT exon 1 as tiny example graph (maybe)

\subsection{Converting to standard text formats}

% pjotr & andrea

% view / paths
Allows us to export basic data of the graph (genomes and graph structure).

Figure: .dot showing the relationship between all the tools and the data formats we use.

\subsection{Visualizing pangenome graphs}

% andrea (viz) & erik (draw)

% interesting that we don't go base-level at all... we use vg for that, or STM

Pangenome graph visualization is one of the first steps to gain insight into the mutual relationship between the sequences and their variation.
\texttt{odgi viz} and \texttt{odgi draw} provide scalable ways of generating human-viewable pictures of the high-level structure of the graph.

\texttt{odgi viz} offers a binned, linearized rendering in 1 dimension (1D).

% Explain odgi viz's visualization

However, complex, nonlinear graph structures are difficult to display and interpret in a convenient low number of dimensions.
To overcome such a limitation, \texttt{odgi viz} supports multiple visualization modes, making it easy to grasp the properties of the graph and its complexity.

% Explain odgi viz's color by path position
% Explain odgi viz's color by strandness
% Explain odgi viz's color by coverage

\texttt{odgi draw} extends the visualization in 2 dimensions (2D), taking in input the layouts built by \texttt{odgi layout}.

% Explain odgi draw's visualization

Figure here with 1D + 2D vizs.

\subsection{Finding latent structure in the pangenome graph topology}

% simon

% sort / layout / NO tension! We keep that for the sorting paper

% Allows us to understand the sparse structures typically found in pangenome graphs.

Pangenome graphs built from raw sets of alignments may have complex structures which can introduce difficulty in downstream analyses. Uncovering their latent topology allows us to understand their typical sparse structures.

\texttt{odgi sort} offers multiple sorting algorithms to find the best node order in one dimension. Most notably, nodes can be sorted topologically, randomly, by breaking cycles in the graph, by grooming, or by using a novel path-guided stochastic gradient descent (PG-SGD) algorithm. 
The latter one exploits the biological information in the paths for sorting the nodes.
Combining a series of sorts can mitigate the weaknesses of individually applied sorting algorithms. 
In Figure \textbf{\textit{TODO MHC wgg.85}}, a comparison of an unsorted human pangenome MHC graph and a sorted one are displayed.

For the establishment of a layout in two dimensions \texttt{odgi layout} applies a 2D adjusted PG-SGD algorithm. A calculated layout can be visualized statically with \texttt{odgi draw} (see Figure \textit{\textbf{TODO MHC wgg.85}}) or interactively explored with the in-development \texttt{gfaestus}\footnote{\url{https://github.com/chfi/gfaestus}}.

\subsection{Obtaining metrics of the pangenome graph}

% simon
% HTT exon1 
% Only chr8 longs if I can't produce someting nice

% stats / multiQC / bin / degree / depth 

% Solves the problem of understanding the overall structure, size, and features of the graph and its sequences.


Graphs statistics provide alternative ways to gain insight into pangenomes complexity revealing the overall structure, size, and features of a graph and its sequences. 

Applying \texttt{odgi stats}, users can retrieve metrics describing the graph properties, such as the number of nodes, edges, paths, and graph length. 
MultiQC's \citep{Ewels_2016} ODGI module\footnote{\url{https://multiqc.info/docs/\#odgi} (accessed Oct 2021)} provides an interactive way to comparatively explore the statistics of an arbitrary number of graphs. 
ODGI also offers more advanced tools for the interrogation of the graphs. \texttt{odgi bin} summarizes the path information into bins of a specified size, enabling a summarized view of gigabase scale graphs. 
This concept found its way into the pangenome graph ontology \cite{Yokoyama2020}. \texttt{odgi depth} returns the node depth as defined by query criteria, allowing users to assess the complexity of regions in the graph. 
Collapsed sequences with highly identical repeats indicate intricate \textit{loci}. 
Complex motifs can be also detected with \texttt{odgi degree}, which returns the node degree as defined by query criteria. 
High degree nodes can be the mirror of variation, misassemblies, or problems in the pangenome building, making the tool useful for validation. 
In Fig. \ref{fig:metrics}, the characteristics of a 90 haplotypes human pangenome graph of the exon 1 huntingtin gene \citep{Sathasivam2013,Neueder2017} (\textit{HTTexon1}), obtained with the ODGI metrics tools, are shown.

\input{fig_metrics}

% Figure: small graph with all the stats shown. A screenshot of MultiQC, visualize summary per bin somehow, R visualizations of degree and depth HTT exon 1 as tiny example graph

\subsection{Editing the graph structure}

% andrea (?)

% chop / unchop / squeeze / break / prune / explode
These are commonly-needed basic operations on the topology of the graph.

\subsection{Untangling and navigating the pangenome with path-based coordinates}

% erik

% position / tips
% untangle (this follows on the jaccard graph mapping concepts in the previous section}
The graph is a model of an alignment of many genomes.
We can use coordinates in any genome to refer to it, but this requires a few basic operations ...
Obtaining unambiguous mappings between different genomes requires the use of a new kind of graph based mapping (path graph jaccard).

Figure showing the untangling of C4 ?
Figure showing tips over a single human chromosome from HPRC.

\subsection{Extracting regions of interest}

% andrea (?)

% extract
Pangenome graphs are very large, but we often only want to work with a small portion (e.g. a single gene).
We can extract such regions using coordinates on the paths in the graph to guide us.

Figure showing extraction process---either schematic or ``real''.

\section{Discussion}

% erik & pjotr & anyone

Pangenome graphs stand to become a ubiquitous tool in genomics \citep{Eizenga_2020}. With ODGI we implemented a state-of-the-art tool suite that can transform, analyse, simplify, validate, and visualise pangenome graphs at large scale. Lifting over annotations and linearizing nested graph structures place the suite as the bridge between traditional linear reference genome analysis and pangenome graphs. ODGI is a unique set of tools that enables scientists to explore and discover the underlying biology of pangenome graphs. Already, the tools are the backbone of pipelines such as the Pangenome Graph Builder\footnote{\url{https://github.com/pangenome/pggb} (accessed Oct 2021)} (PGGB) or nf-core/pangenome\footnote{\url{https://github.com/nf-core/pangenome} (accessed Oct 2021)}. Future work will add support for RNA and protein sequences and expand on metadata capabilities of large pangenome graphs.

rdf sparql side, bring in annotation for great biological interpretation \cite{Yokoyama2020}, federated queries, sharing, FAIR

(recent) interactive visualization tools?! CITE PANACHE, also indicate that no published tool for interactive exploration of such graphs exist

some cancer stuff?!

\section*{Funding}

S.H. acknowledges funding from the Central Innovation Programme (ZIM) for SMEs of the Federal Ministry for Economic Affairs and Energy of Germany.
This work was supported by the BMBF-funded de.NBI Cloud within the German Network for Bioinformatics Infrastructure (de.NBI) (031A537B, 031A533A, 031A538A, 031A533B, 031A535A, 031A537C, 031A534A, 031A532B).

\section*{Data availability}

Data used to build Human pangenome graphs is available at \url{https://github.com/human-pangenomics/HPP_Year1_Data_Freeze_v1.0}.
TODO!

\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
%
%\bibliographystyle{plain}
%
\bibliography{document}


% \begin{thebibliography}{}

% \bibitem[Bofelli {\it et~al}., 2000]{Boffelli03}
% Bofelli,F., Name2, Name3 (2003) Article title, {\it Journal Name}, {\bf 199}, 133-154.

% \bibitem[Bag {\it et~al}., 2001]{Bag01}
% Bag,M., Name2, Name3 (2001) Article title, {\it Journal Name}, {\bf 99}, 33-54.

% \bibitem[Yoo \textit{et~al}., 2003]{Yoo03}
% Yoo,M.S. \textit{et~al}. (2003) Oxidative stress regulated genes
% in nigral dopaminergic neurnol cell: correlation with the known
% pathology in Parkinson's disease. \textit{Brain Res. Mol. Brain
% Res.}, \textbf{110}(Suppl. 1), 76--84.

% \bibitem[Lehmann, 1986]{Leh86}
% Lehmann,E.L. (1986) Chapter title. \textit{Book Title}. Vol.~1, 2nd edn. Springer-Verlag, New York.

% \bibitem[Crenshaw and Jones, 2003]{Cre03}
% Crenshaw, B.,III, and Jones, W.B.,Jr (2003) The future of clinical
% cancer management: one tumor, one chip. \textit{Bioinformatics},
% doi:10.1093/bioinformatics/btn000.

% \bibitem[Auhtor \textit{et~al}. (2000)]{Aut00}
% Auhtor,A.B. \textit{et~al}. (2000) Chapter title. In Smith, A.C.
% (ed.), \textit{Book Title}, 2nd edn. Publisher, Location, Vol. 1, pp.
% ???--???.

% \bibitem[Bardet, 1920]{Bar20}
% Bardet, G. (1920) Sur un syndrome d'obesite infantile avec
% polydactylie et retinite pigmentaire (contribution a l'etude des
% formes cliniques de l'obesite hypophysaire). PhD Thesis, name of
% institution, Paris, France.

% \end{thebibliography}
\end{document}
