\documentclass{bioinfo}

\usepackage{relsize} % for Cpp
\usepackage{xspace} % for Cpp
\usepackage{subfigure}

\usepackage{natbib}

\newcommand{\odgi}{ODGI}
\newcommand{\vg}{VG}

\newcommand{\FIXME}{\textbf{!!FIXME!!}}
\newcommand{\cmd}[1]{\textrm{#1}}
\newcommand{\cmdbf}[1]{{\textbf{#1}}}
\newcommand{\topic}[1]{{\cmdbf{#1}}:}
\newcommand{\Rplus}{\protect\hspace{-.1em}\protect\raisebox{.35ex}{\smaller{\smaller\textbf{+}}}}
\newcommand{\pp}{\Rplus\Rplus}
\newcommand{\Cpp}{\mbox{C\Rplus\Rplus}\xspace}


\copyrightyear{2021} \pubyear{2021}

\access{Advance Access Publication Date: Day Month Year}
\appnotes{Application Note}

\begin{document}
    \firstpage{1}

    \subtitle{Pangenome analysis}

    \title[ODGI: scalable tools for pangenome graphs]{ODGI: scalable tools for pangenome graphs}
    \author[Guarracino, Heumos \textit{et~al}.]{
        Andrea Guarracino\,$^{\text{\sfb 1,}*}$,
        Simon Heumos\,$^{\text{\sfb 2,}*}$,
    % XXXXX,
        Pjotr~Prins\,$^{\text{\sfb 3}}$ and
        Erik~Garrison\,$^{\text{\sfb 3,}\dagger}$
        \footnote{The authors wish it to be known that the first two authors should be regarded as joint first authors.}
    }
    \address{$^{\text{\sf 1}}$Biology, University of Tor Vergata, Rome, 00133, Italy \\
        $^{\text{\sf 2}}$Quantitative Biology Center (QBiC), University of Tübingen, Tübingen, 72076, Germany \\
        $^{\text{\sf 3}}$Dept. Genetics Genomics \& Informatics, UTHSC, Memphis, TN, USA
    }

    \corresp{$^\dagger$To whom correspondence should be addressed.}

    \history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

    \editor{Associate Editor: XXXXXXX}

    \abstract{ \textbf{Motivation:} Pangenomes graphs address some of the
      shortcomings of mainline genomics, in particular
      with regard to reference genome bias and identification of complex
      structural variants.
        %
        \\ \textbf{Results:} Here we present \odgi\, a novel range of
        scalable tools that implements new algorithms and has an efficient in-memory
        representation of DNA graphs. \odgi\ can handle hundreds of
        large genomes in memory with fast parallel execution. \odgi\ includes
        tools for detecting complex regions, extracting \textit{loci}, removing artifacts,
        exploratory analysis, and graph sorting, manipulation and visualisation.
        \odgi\ also provides an extensible API that allows for writing new tool.
        %
        \\
        \textbf{Availability:} \odgi\ is software written in the \Cpp\ programming language and
        published under the permissive MIT license. Source code and documentation can be downloaded from
        \href{https://github.com/pangenome/odgi}{https://github.com/pangenome/odgi}.\\
        %
        \textbf{Contact:}
        % \href{andreaguarracino@outlook.com}{andreaguarracino@outlook.com},
        egarris5@uthcs.edu
        \\
        \textbf{Supplementary information:} Supplementary data are available at \textit{Bioinformatics} online. \FIXME
        }
    \maketitle


    \section{Introduction}

    Thanks to advances in sequencing technology, new full-length
    genome assemblies are produced at a high rate, including efforts
    from the Human Pangenome Reference Consortium (HPRC) and
    telomere-to-telemere projects, e.g.~\citep{Miga:2020, Logsdon:2021}. In
    principle, a pangenome can model a full set of genomic elements of
    individuals in a given species or clade. Pangenomics contrasts
    with reference genome based approaches which relate sequences to a
    single consensus model of the genome~\citep{Eizenga:2020}. A
    pangenome can elegantly be represented by a graph data structure
    incorporating sequences as nodes and their connections as
    edges. These nodes are shared for identical sequences, such as
    homologs, paralogs and orthologs. So-called paths can represent
    reads, contigs, haplotypes, individual chromosomes or part
    thereof, that run through the nodes. A bi-directed graph can even
    contain both strands of DNA as paths. A graph-based approach is
    particularly relevant for exploring complex variation and highly
    repetitive regions, such as the Major Histocompatibility Complex
    (MHC, Fig.~\ref{fig:1}).

    A major challenge is writing software that can deal with the sheer
    size of these graphs representing hundreds, and soon thousands, of
    human genomes. The \vg\ toolkit pioneered graph
    processing~\citep{vgtools,Eizenga:2020b}, but we found it does not
    necessarily scale for large data (see Table~\ref{tab:02}). We
    wrote a new toolset and Application Programming Interface (API) in
    \Cpp\ named Optimized Dynamic Genome/Graph Implementation (ODGI).
    \odgi\ implements a efficient graph structure in computer memory
    that can be dynamically updated using multiple CPU cores in
    parallel.

    Note that \odgi\ is focused on `first-order' graph
    operations. Other tools are required to construct graphs from raw
    sequences, map reads, or call variants. \odgi\ provides
    operations on an existing graph, such as subset, subdivide, break,
    combine, normalize and order its components, nodes, and
    paths. Most of the tools are designed to be applied together,
    piping the output from one tool into the next.


    \input{figure_p}

    \vspace{-0.15in}

    \section{Results}

    \odgi\ is designed to build and modify paths in parallel. It
    applies concepts first introduced in the dynamic version of the
    graph extension of the positional Burrows-Wheeler
    transform~\citep{Siren:2020}. For implementation details see the
    \odgi\ online documentation. \odgi\ currently comes with more than 30
    tools and here we present a subset of functionality:

    \indent \topic{odgi build, view \& validate} pangenome graphs can be
    stored in the textual Graphical Fragment Assembly
    format~\citep{GFA}. \cmd{odgi build} and \cmd{odgi view}
    convert graphs from GFA to the \odgi\ format and \textit{vice versa}.
    \cmd{odgi validate} ascertains the graph data is correct.

    \topic{odgi viz, sort, draw \& layout} pangenome visualisation
    provides convenient insight into genomic variation. \cmd{odgi viz}
    generates a linearized representation of the pangenome (see Fig.~\ref{fig:1}a)
    and is capable of handling full length human chromosomes. \cmd{odgi sort}
    reorganises node order and simplifies the graph by applying several
    algorithms, including our novel 'path-guided stochastic gradient descent'
    algorithm. \cmd{odgi layout} generates a layout file with the X-Y
    coordinates for all nodes and \cmd{odgi draw} creates the figure.

    \topic{odgi stats, bin, depth \& degree} pangenome numerical
    counts provide insight into complexity in other
    ways. \cmd{odgi stats} returns the number of nodes,
    edges, paths, and graph length. \cmd{odgi bin} summarises path
    information into bins of a specified size, thereby enabling a
    summarised view of large genomes. \cmd{odgi depth} and
    \cmd{odgi degree} compute node depth and degree as defined by
    user provided criteria. These methods allow the detection of complex
    genomic regions generated by highly repetitive sequences.

    \topic{odgi break, groom, chop \& unchop} pangenome graphs can be
    reorganised. \cmd{odgi break} removes circular structures in the
    graph, thereby reducing its complexity. \cmd{odgi groom} removes
    spurious inverted links by aligning the graph from the orientation
    that is supported by most paths. \cmd{odgi chop} divides long
    nodes into shorter ones at a maximum requested size, thereby
    simplifying downstream analysis. \cmd{odgi unchop} joins nodes
    and embedded sequences that do not change the graph topology,
    thereby obtaining a more compact representation of the graph.


    \topic{odgi explode, squeeze \& extract} pangenomes are constructed
    as large graphs. \cmd{odgi explode} separates units, such as
    chromosomes, into different files. \cmd{odgi squeeze} merges
    multiple graphs into the same file whilst preventing node ID
    collisions. \cmd{odgi extract} extracts regions of the graph as
    defined by certain criteria, allowing downstream processing of
    smaller subgraphs.

    \topic{odgi position} pangenome graphs are flexible when it comes
    to coordinate systems. \cmd{odgi position} can use the coordinate
    system from a contained reference genome --- a dynamic liftover
    --- to display coordinates and other localised features, as is shown in
    Fig.~\ref{fig:1b}. Note that multiple reference genomes can be contained
    in the graph and any contained path can be used as a reference.

    \vspace{-0.1in}

    \section{Discussion}

    Pangenome graphs stand to become a ubiquitous tool in
    genomics~\citep{Eizenga:2020}. With \odgi\ we implemented a
    state-of-the-art tool suite that can transform, analyse and
    visualise pangenome graphs at large scale. \odgi\ also exposes an
    extensible API that can be used by other tools, e.g., for
    genotyping, pangenome graph building and normalization.  Future
    work will add support for RNA and protein sequences and expand on
    metadata capabilities of large pangenome graphs.

    \input{table}


%\begin{figure}[!tpb]%figure2
%%\centerline{\includegraphics{fig02.eps}}
%\caption{Caption, caption.}\label{fig:02}
%\end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%     please remove the " % " symbol from \centerline{\includegraphics{fig01.eps}}
%     as it may ignore the figures.
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

    \textit{Financial Support}: \FIXME\ Garrison \& Prins.
    % \vspace*{+12pt}

    \textit{Conflict of Interest}: none declared.

    % \vspace*{+12pt}


    % \section*{Data availability}

    % Data used to build Human pangenome graphs is available at \href{https://github.com/human-pangenomics/HPP_Year1_Data_Freeze_v1.0}{https://github.com/human-pangenomics/HPP\_Year1\_Data\_Freeze\_v1.0}.

    % \section*{Acknowledgements}
    %ToDo
    % XXXXXXX XXXXXXXX XXXXXXXX XXXXXXXX XXXXXXXX XXXXXXXX XXXXXXXX XXXXXXXX XXXXXXXX.


    % \section*{Funding}
    %ToDo
    % This work has been supported by the XXXXXXXX XXXXXXXX XXXXXXXX XXXXXXXX XXXXXXXX.

%    \vspace*{-12pt}
    \vspace{-0.15in}

\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
%
%\bibliographystyle{plain}
%
%\bibliography{Document}

\bibliography{bibliography}

\end{document}
