\PassOptionsToPackage{utf8}{inputenc}
\documentclass{bioinfo}

\usepackage[draft]{hyperref}
\usepackage{makecell}

\newcommand{\vocab}{\textbf}

\copyrightyear{XXXX} \pubyear{XXXX}

\access{Advance Access Publication Date: Day Month Year}
\appnotes{Genome Analysis}

\begin{document}
\firstpage{1}

\subtitle{Genome Analysis}

\title[short Title]{ODGI: scalable tools for working with pangenome graphs}
\author[Heumos, \textit{et~al}.]{
Simon Heumos\,$^{\text{\sfb 1, \sfb 2} \dagger}$,
Andrea Guarracino\,$^{\text{\sfb 2} \dagger}$,
Pjotr Prins\,$^{\text{\sfb 3}}$,
and Erik Garrison\,$^{\text{\sfb 3}*}$
}

\address{
$^{\text{\sf 1}}$Quantitative Biology Center (QBiC), University of T\"ubingen, T\"ubingen, Germany, 72076
$^{\text{\sf 2}}$University of Tor Vergata, Rome, Italy and \\
$^{\text{\sf 3}}$University of Tennessee Health Science Center, Memphis, TN, USA
}

\corresp{$^\ast$To whom correspondence should be addressed. \\
$^\dagger$Contributed equally.}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\abstract{
\textbf{Motivation:} Pangenome graphs provide a complete representation of the mutual alignment of collections of genomes.
These models, therefore, offer the opportunity to study the entire genetic diversity of a population, bringing new challenges to how to analyze hundreds of genomes on a gigabase scale.
In addition, vertebrate genomes present highly repetitive regions which increase the complexity of the analysis.
Efficient and flexible tools are therefore required to work on pangenome graphs of any complexity and scale. \\
\textbf{Results:} Pangenome graphs provide a complete representation of the mutual alignment of collections of genomes.
These models, therefore, offer the opportunity to study the entire genetic diversity of a population, bringing new challenges to how to analyze hundreds of genomes on a gigabase scale.
In addition, vertebrate genomes present highly repetitive regions which increase the complexity of the analysis.
Efficient and flexible tools are therefore required to work on pangenome graphs of any complexity and scale. \\
\textbf{Availability:} ODGI is written in C++. Its source code is freely available at \url{https://github.com/pangenome/odgi} with documentation at \url{https://odgi.readthedocs.io}.
ODGI can be installed via Bioconda \url{https://bioconda.github.io/recipes/odgi/README.html} or Guix https://github.com/ekg/guix-genomics/blob/master/odgi.scm. \\
\textbf{Contact:} \href{egarris5@uthsc.edu}{egarris5@uthsc.edu} \\
\textbf{Supplementary information:} Supplementary data are available at \textit{Bioinformatics} online.
}

\maketitle

\section{Introduction}

\cite{Eizenga_2020}

\begin{methods}

\section{Implementation}

\subsection{Data model}

\subsection{The \textsc{HandleGraph} interface}

\subsection{Code availability}

\end{methods}

\section{Evaluation}

\subsection{Human genome with structural variants}

\subsection{Genome graph collection}

\subsection{1000 Genome Project chromosome graphs}

\section{Discussion}

\section*{Funding}

%This work was supported, in part, by the National Institutes of Health (award numbers U01HG010961, U41HG010972, R01HG010485, 2U41HG007234, 5U54HG007990, 5T32HG008345-04, U01HL137183 to B.P.) and the W. M. Keck Foundation (award number DT06172015 to B.P.).
S.H. acknowledges funding from the Central Innovation Programme (ZIM) for SMEs of the Federal Ministry for Economic Affairs and Energy of Germany.

\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
%
%\bibliographystyle{plain}
%
\bibliography{document}


% \begin{thebibliography}{}

% \bibitem[Bofelli {\it et~al}., 2000]{Boffelli03}
% Bofelli,F., Name2, Name3 (2003) Article title, {\it Journal Name}, {\bf 199}, 133-154.

% \bibitem[Bag {\it et~al}., 2001]{Bag01}
% Bag,M., Name2, Name3 (2001) Article title, {\it Journal Name}, {\bf 99}, 33-54.

% \bibitem[Yoo \textit{et~al}., 2003]{Yoo03}
% Yoo,M.S. \textit{et~al}. (2003) Oxidative stress regulated genes
% in nigral dopaminergic neurnol cell: correlation with the known
% pathology in Parkinson's disease. \textit{Brain Res. Mol. Brain
% Res.}, \textbf{110}(Suppl. 1), 76--84.

% \bibitem[Lehmann, 1986]{Leh86}
% Lehmann,E.L. (1986) Chapter title. \textit{Book Title}. Vol.~1, 2nd edn. Springer-Verlag, New York.

% \bibitem[Crenshaw and Jones, 2003]{Cre03}
% Crenshaw, B.,III, and Jones, W.B.,Jr (2003) The future of clinical
% cancer management: one tumor, one chip. \textit{Bioinformatics},
% doi:10.1093/bioinformatics/btn000.

% \bibitem[Auhtor \textit{et~al}. (2000)]{Aut00}
% Auhtor,A.B. \textit{et~al}. (2000) Chapter title. In Smith, A.C.
% (ed.), \textit{Book Title}, 2nd edn. Publisher, Location, Vol. 1, pp.
% ???--???.

% \bibitem[Bardet, 1920]{Bar20}
% Bardet, G. (1920) Sur un syndrome d'obesite infantile avec
% polydactylie et retinite pigmentaire (contribution a l'etude des
% formes cliniques de l'obesite hypophysaire). PhD Thesis, name of
% institution, Paris, France.

% \end{thebibliography}
\end{document}
